File: frontend/src/types/portfolio.ts
/**
 * Portfolio type definitions that:
 * - Define holding and transaction structures
 * - Specify portfolio metrics and summary interfaces
 * - Manage historical data point types
 * - Handle portfolio state management types
 */

// src/types/portfolio.ts
export interface Holding {
  id: string;
  symbol: string;
  name: string;
  shares: number;
  avg_cost: number;
  current_price: number;
  market_value: number;
  day_change: number;
  day_change_percent: number;
  total_return: number;
  total_return_percent: number;
  weight: number;
  sector: string;
  industry: string;
  last_price_update: string;
}

export interface Transaction {
  id: string;
  symbol: string;
  type: 'BUY' | 'SELL';
  shares: number;
  price: number;
  amount: number;
  fees: number;
  date: string;
  note?: string;
}

export interface PortfolioSummary {
  total_value: number;
  total_cost: number;
  total_return: number;
  total_return_percent: number;
  day_change: number;
  day_change_percent: number;
  cash_balance: number;
  dividend_income: number;
  last_sync: string;
}

export interface PortfolioMetrics {
  sharpeRatio: number;
  beta: number;
  alpha: number;
  var95: number;
  maxDrawdown: number;
  volatility: number;
  annualizedReturn: number;
  informationRatio: number;
  treynorRatio: number;
  calmarRatio: number;
}

export interface HistoricalDataPoint {
  date: string;
  value: number;
  change: number;
  changePercent: number;
}

export interface SectorAllocation {
  sector: string;
  value: number;
  percentage: number;
  count: number;
}

export interface PortfolioState {
  summary: PortfolioSummary | null;
  holdings: Holding[];
  transactions: Transaction[];
  metrics: PortfolioMetrics | null;
  history: HistoricalDataPoint[];
  sectorAllocation: SectorAllocation[];
  isLoading: boolean;
  error: string | null;
}

File: backend/portfolio_service/app/schemas/portfolio.py
# app/schemas/portfolio.py
from pydantic import BaseModel, validator
from datetime import datetime
from typing import List, Optional

class PortfolioBase(BaseModel):
    name: str
    description: Optional[str] = None

class PortfolioCreate(PortfolioBase):
    pass

class PortfolioUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None

class PortfolioSummary(BaseModel):
    total_value: float
    total_cost: float
    total_return: float
    total_return_percent: float
    day_change: float
    day_change_percent: float
    cash_balance: float
    dividend_income: float
    last_sync: Optional[datetime]

class PortfolioMetrics(BaseModel):
    sharpe_ratio: float
    beta: float
    alpha: float
    var_95: float
    max_drawdown: float
    volatility: float
    annualized_return: float
    information_ratio: float

class PortfolioResponse(PortfolioBase):
    id: int
    user_id: int
    summary: PortfolioSummary
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True

# app/schemas/holding.py
class HoldingBase(BaseModel):
    symbol: str
    shares: float
    avg_cost: float

class HoldingCreate(HoldingBase):
    pass

class HoldingUpdate(BaseModel):
    shares: Optional[float] = None
    avg_cost: Optional[float] = None

class HoldingResponse(HoldingBase):
    id: int
    name: str
    current_price: float
    market_value: float
    day_change: float
    day_change_percent: float
    total_return: float
    total_return_percent: float
    weight: float
    sector: Optional[str]
    industry: Optional[str]
    last_price_update: Optional[datetime]
    
    class Config:
        from_attributes = True

# app/schemas/transaction.py
class TransactionBase(BaseModel):
    symbol: str
    type: str
    shares: float
    price: float
    fees: float = 0.0
    note: Optional[str] = None

class TransactionCreate(TransactionBase):
    transaction_date: Optional[datetime] = None
    
    @validator('type')
    def validate_type(cls, v):
        if v not in ['BUY', 'SELL', 'DIVIDEND']:
            raise ValueError('Type must be BUY, SELL, or DIVIDEND')
        return v

class TransactionResponse(TransactionBase):
    id: int
    total_amount: float
    transaction_date: datetime
    created_at: datetime
    
    class Config:
        from_attributes = True

# Portfolio data validation schemas that:
# - Define data structures for portfolio operations
# - Validate portfolio creation and updates
# - Structure portfolio summary responses
# - Define portfolio metrics format
